README.md
# SQL Driver for Cockpit CMS

[![Latest Version](https://img.shields.io/packagist/v/piotr-cz/cockpit-sql-driver?style=flat-square&sort=semver)](https://packagist.org/packages/piotr-cz/cockpit-sql-driver)
[![Build status](https://img.shields.io/travis/piotr-cz/cockpit-sql-driver?style=flat-square)](https://travis-ci.org/piotr-cz/cockpit-sql-driver)

This addon allows to use MySQL/ MariaDB/ PostgreSQL databases instead of default MongoDB/ SQLite.


## Requirements

- Cockpit CMS (next, tested up to v0.11.0 or legacy)
- MySQL 5.7.9/ MariaDB 10.2.6/ PostgreSQL 9.4
- PHP 7.1
- Enabled PHP extensions: *pdo*, *pdo_mysql*/ *pdo_pgsql*


### Compatibility

To run Cockpit `0.9.2+` at least version `1.0.0-beta.2+` is required.


## Installation

_Note:_

If you installed addon before ever starting Cockpit, [some errors](#error-call-to-a-member-function-toarray-on-null) may come up once you start it.

To solve it, start Cockpit with database configuration it supports [out of the box](https://getcockpit.com/documentation/reference/configuration) to trigger Cockpit warmup and then set [configuration](#configuration) specific for this addon.


### Manual

Download [latest release](https://github.com/piotr-cz/cockpit-sql-driver/releases/latest) and extract to `COCKPIT_PATH/addons/SqlDriver` directory


### Using Cockpit CLI _(development version)_

```sh
./cp install/addon --name SqlDriver --url https://github.com/piotr-cz/cockpit-sql-driver/archive/master.zip
```


### Using Composer

1. Make sure path to cockpit addons are defined in your projects' _composer.json_ file

   ```json
   {
       "name": "MY_PROJECT",
       "extra": {
           "installer-paths": {
               "public/cockpit/addons/{$name}": ["type:cockpit-module"]
           }
       }
   }
   ```

2. In your project root run command

   ```sh
   composer require piotr-cz/cockpit-sql-driver
   ```


## Configuration

Example configuration for `COCKPIT_PATH/config/config.php` file:

```php
<?php
return [
    # Cockpit configuration
    # …

    # Use SQL Driver as main data storage
    'database' => [
        'server' => 'sqldriver',
        # Connection options
        'options' => [
            'connection' => 'mysql',         # One of 'mysql'|'pgsql'
            'host'       => 'localhost',     # Optional, defaults to 'localhost'
            'port'       => 3306,            # Optional, defaults to 3306 (MySQL) or 5432 (PostgreSQL)
            'dbname'     => 'DATABASE_NAME',
            'username'   => 'USER',
            'password'   => 'PASSWORD',
            'charset'    => 'UTF8',          # Optional, defaults to 'UTF8'
            'tablePrefix' => '',             # Optional, database tables prefix (ie. 'cockpit_')
            'bootstrapPriority' => 999,      # Optional, defaults to 999
        ],
        # Connection specific options
        # General: https://www.php.net/manual/en/pdo.setattribute.php
        # MySQL specific: https://www.php.net/manual/en/ref.pdo-mysql.php#pdo-mysql.constants
        'driverOptions' => [],
    ],
];
```

_Rererence: Cockpit docs > [Configuration](https://getcockpit.com/documentation/reference/configuration)_


## Database content migration (Cockpit v0.6.0+)

1. Export data to `COCKPIT_PATH/migration` subdirectory

   ```sh
   mkdir migration
   ./cp export --target migration
   ```

2. Switch database to _sqldriver_ (see [Configuration](#configuration))

3. Import data from `COCKPIT_PATH/migration` subdirectory

   ```sh
   ./cp import --src migration
   rm -rf migration
   ```

_Reference: Cockpit docs > [CLI](https://getcockpit.com/documentation/reference/CLI)_


## Testing

There are integration tests included in the package.
These require Cockpit CMS as a dev dependency and use it's _MongoHybrid\Client_ API to run actions on database.

To run tests

1. Install dependencies

   ```sh
   cd COCKPIT_PATH/addons/SqlDriver
   composer install
   ```

2. Configure test database

   copy [`/phpunit.xml.dist`](./phpunit.xml.dist) to `/phpunit.xml` and set up variables as in [configuration](#configuration)

3. Run tests with PHPUnit

   ```sh
   ./vendor/bin/phpunit
   ```


## Drawbacks

Cockpit doesn't provide public API to register custom database drivers so this addon monkey-patches Cockpit Driver selector client (_MongoHybrid Client_).
This means that there is no guarantee that this addon will work in future versions of Cockpit.


### Collection filters

#### Not implemented

- `$func`/ `$fn`/ `$f`

- `$fuzzy`


#### Work differently

- _callable_

  [Unlike SQLite](https://www.php.net/manual/en/pdo.sqlitecreatefunction.php), PDO MySQL and PostgreSQL drivers don't have support for User Defined Functions in PHP language - so callable is evaluated on every result fetch.
  If you have lots of documents in collection and care about performance use other filters.

- `$in`, `$nin`

  When database value is an array, evaluates to false.

- `$regexp`
  - MySQL implemented via [REGEXP](https://dev.mysql.com/doc/refman/5.7/en/regexp.html) + case insensitive
  - PostgreSQL impemeted via [POSIX Regular Expressions](https://www.postgresql.org/docs/9.4/functions-matching.html#FUNCTIONS-POSIX-REGEXP) + case insensitive

  Wrapping expression in `//` or adding flags like `/foobar/i` won't work, as MySQL and PosgreSQL Regexp functions don't support flags.

- `$text`
  - MySQL implemeted via [LIKE](https://dev.mysql.com/doc/refman/5.7/en/string-comparison-functions.html#operator_like)
  - PostgreSQL implementad via [LIKE](https://www.postgresql.org/docs/9.4/functions-matching.html#FUNCTIONS-LIKE)

  Filter options are not supported (_$minScore_, _$distance_, _$search_).


## Manual database optimizations

By default package creates virtual column `_id` with unique index on every created collection.

If you would like to speed up filters on other collection fields - add virtual column with suitable index and type.

For example to add virtual column of integer type for field FIELD_NAME in TABLE_NAME collection, use

- MySQL:

  ```sql
  ALTER TABLE
      `{TABLE_NAME}` ADD COLUMN `{FIELD_NAME}_virtual` INT AS (`document` ->> '$.{FIELD_NAME}') NOT NULL,
      ADD UNIQUE | KEY `idx_{TABLE_NAME}_{FIELD_NAME}` (`{FIELD_NAME}_virtual`);
  ```

  _Reference: MySQL 5.7 > [CREATE INDEX](https://dev.mysql.com/doc/refman/5.7/en/create-index.html)_

- PosgreSQL:

  ```sql
  CREATE [UNIQUE] INDEX "idx_{TABLE_NAME}_{FIELD_NAME}" ON "{FIELD_NAME}" ((("document" ->> '{FIELD_NAME}')::int));
  ```

  _Reference: PostgreSQL 9.4 > [CREATE INDEX](https://www.postgresql.org/docs/9.4/sql-createindex.html)_


## Known issues

### Error: `Call to a member function toArray() on null`

This happens when starting cockpit for the first time and this addon is installed.
The reason is in that native Cockpit modules try to accesss storage which is initialized later (during custom modules bootstrap).

Cockpit must be started for the first time without being configured to use SQL driver.

**Solution 1**
Start Cockpit with database configuration it supports out of the box and than switch to `sqldriver` as described [here](#configuration)

**Solution 2**
Manually create file `COCKPIT_STORAGE_FOLDER/tmp/webhooks.cache.php` with content

```php
<?php return [];
```


## Copyright and license

Copyright since 2019 Piotr Konieczny under the MIT license.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2021-04-27
### Fixed
- PHP 8.0 compatibility (#fdc1f5c)
- Fix MariaDB version detection ([#8](https://github.com/piotr-cz/cockpit-sql-driver/pull/8))

## [1.0.0-rc.3] - 2020-08-07
### Added
- Table prefix option

### Changed
- Change interfaces implementations

### Fixed
- Increase default addon boostrap priority

## [1.0.0-rc.2] - 2020-07-15
### Fixed
- Use Driver::insert for one or many documents, change method signatures
  Fixes cockpit assets integration

## [1.0.0-rc.1] - 2020-03-05
### Fixed
- Allow nulls booleans in filter
- Differences in JSON path selector in MySQL

## [1.0.0-beta.2] - 2020-02-21
### Added
- Added `Collection::insertMany` method (fixes compatibility with Cockpit v0.9.3+)

## [1.0.0-beta.1] - 2019-09-16
### Added
- Initial release

[Unreleased]: https://github.com/piotr-cz/cockpit-sql-driver/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/piotr-cz/cockpit-sql-driver/compare/v1.0.0-rc.3...v1.0.0
[1.0.0-rc.3]: https://github.com/piotr-cz/cockpit-sql-driver/compare/v1.0.0-rc.2...v1.0.0-rc.3
[1.0.0-rc.2]: https://github.com/piotr-cz/cockpit-sql-driver/compare/v1.0.0-rc.1...v1.0.0-rc.2
[1.0.0-rc.1]: https://github.com/piotr-cz/cockpit-sql-driver/compare/v1.0.0-beta.2...v1.0.0-rc.1
[1.0.0-beta.2]: https://github.com/piotr-cz/cockpit-sql-driver/compare/v1.0.0-beta.1...v1.0.0-beta.2
[1.0.0-beta.1]: https://github.com/piotr-cz/cockpit-sql-driver/releases/tag/v1.0.0-beta.1


lib/MongoSql/Driver/MysqlDriver.php
<?php
declare(strict_types=1);

namespace MongoSql\Driver;

use PDO;

use MongoSql\Driver\Driver;

/**
 * Cockpit CMS MySQL Driver
 * Requires MySQL 5.7.9+ (JSON support and shorthand operators)
 *          or MariaDB 10.2.3+ (JSON support: 10.2.3, Generated columns: 10.2.6)
 */
class MysqlDriver extends Driver
{
    /** @inheritdoc */
    protected const DB_DRIVER_NAME = 'mysql';

    /** @var string - Min db server version */
    protected const DB_MIN_SERVER_VERSION = '5.7.9';
    protected const DB_MIN_SERVER_VERSION_MARIADB = '10.2.6';

    /**
     * @inheritdoc
     */
    protected static function createConnection(array $options, array $driverOptions = []): PDO
    {
        // See https://www.php.net/manual/en/ref.pdo-mysql.connection.php
        $connection = new PDO(
            vsprintf('mysql:host=%s;port=%s;dbname=%s;charset=%s', [
                $options['host'] ?? 'localhost',
                $options['port'] ?? 3306,
                $options['dbname'],
                $options['charset'] ?? 'UTF8'
            ]),
            $options['username'],
            $options['password'],
            $driverOptions + [
                // Set UTF-8 as character set and collation (Note: Setting sql_mode doesn't work in init command, at least in 5.7.26)
                PDO::MYSQL_ATTR_INIT_COMMAND => 'SET NAMES utf8mb4 COLLATE utf8mb4_unicode_ci;',
                // Use unbuffered query to get results one by one
                PDO::MYSQL_ATTR_USE_BUFFERED_QUERY => false,
            ]
        );

        /* Set sql_mode after connection has started to ISO/IEC 9075
         * https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html
         * https://mariadb.com/kb/en/library/sql-mode/
         */
        $connection->exec("SET sql_mode = 'ANSI';");

        return $connection;
    }

    /**
     * @inheritdoc
     *
     * Version string examples:
     * - MySQL: `5.7.27-0ubuntu0.18.04.1`
     * - MariaDB: `5.5.5-10.2.26-MariaDB-1:10.2.26+maria~bionic`
     * - MariaDB: `5.5.5-10.4.18-MariaDB-cll-lve`
     * - MariaDB: `10.4.18-MariaDB-cll-lve`
     */
    protected function assertIsDbSupported(): void
    {
        parent::assertIsDbSupported();

        $fullVersion = $this->connection->getAttribute(PDO::ATTR_SERVER_VERSION);
        $fullVersionFragments = explode('-', $fullVersion);

        $currentVersion = array_shift($fullVersionFragments);
        $minVersion = static::DB_MIN_SERVER_VERSION;

        // Detect MariaDB
        if (in_array('MariaDB', $fullVersionFragments)) {
            // Detect MySQL compat prefix (replication version hack) if present as first fragment
            // Note: Not present when using `SELECT VERSION()` query
            if ($currentVersion === '5.5.5') {
                $currentVersion = array_shift($fullVersionFragments);
            }

            $minVersion = static::DB_MIN_SERVER_VERSION_MARIADB;
        }

        static::assertIsDbVersionSupported(
            $currentVersion,
            $minVersion
        );
    }
}
